src/main/java/org/spongepowered/api/text/sink/TransformingSink.java

```diff
@@ -0,0 +1,51 @@
+/*
+ * This file is part of SpongeAPI, licensed under the MIT License (MIT).
+ *
+ * Copyright (c) SpongePowered <https://www.spongepowered.org>
+ * Copyright (c) contributors
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package org.spongepowered.api.text.sink;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+import org.spongepowered.api.command.CommandSource;
+import org.spongepowered.api.text.Text;
+
+import java.util.Optional;
+import java.util.function.BiFunction;
+
+public class TransformingSink extends ForwardingSink {
+
+    protected final BiFunction<CommandSource, Optional<Text>, Optional<Text>> function;
+
+    public TransformingSink(MessageSink delegate, BiFunction<CommandSource, Optional<Text>, Optional<Text>> function) {
+        super(delegate);
+        this.function = checkNotNull(function, "function");
+    }
+
+    @Override
+    public Optional<Text> transformMessage(CommandSource recipient, Text original) {
+        Optional<Text> text = this.delegate.transformMessage(recipient, original);
+
+        return this.function.apply(recipient, text);
```

Comment:
It returns an `Optional<Text>`, I'm not sure a null check is really required here (it isn't anywhere else).
It accepts, and returns, `Optional<Text>` so that nulls are not passed around.